tests/system/test_access.py

```diff
@@ -16,44 +20,316 @@
         assert r.status_code == 202, r.status_code
 
-        r = requests.post(url,
-                          data=events,
-                          headers=oauth('Bearer wrongtoken'))
-        assert r.status_code == 401, r.status_code
 
-        r = requests.post(url,
-                          data=events,
-                          headers=oauth('Wrongbearer 1234'))
-        assert r.status_code == 401, r.status_code
+@integration_test
+class BaseAPIKeySetup(ServerBaseTest):
+
+    @classmethod
+    def setUpClass(cls):
+        # According to https://docs.python.org/2/library/unittest.html#setupclass-and-teardownclass setUp and tearDown
+        # should be skipped when class is skipped, which is apparently not true.
+        # This is a hack to avoid running the setup while it should be skipped
+        if not INTEGRATION_TESTS:
+            return
+
+        # application
+        cls.application = "apm"
+
+        # apm-backend privileges
+        cls.privilege_intake = "event:write"
+        cls.privilege_config = "config_agent:read"
+        cls.privilege_sourcemap = "sourcemap:write"
+        cls.privileges_all = [cls.privilege_intake, cls.privilege_sourcemap, cls.privilege_config]
+        cls.privilege_any = "*"
+
+        # resources
+        cls.resource_any = ["*"]
+        cls.resource_backend = ["-"]
+
+        user = os.getenv("ES_SUPERUSER_USER", "admin")
+        password = os.getenv("ES_SUPERUSER_PASS", "changeme")
+        cls.admin_es_url = cls.get_elasticsearch_url(user, password)
+
+        content_type = 'application/json'
+
+        # create privileges
+        url_privileges = "{}/_security/privilege".format(cls.admin_es_url)
+        payload = json.dumps({cls.application: {
+            "sourcemap": {"actions": [cls.privilege_sourcemap]},
+            "intake": {"actions": [cls.privilege_intake]},
+            "config": {"actions": [cls.privilege_config]}}})
+        resp = requests.put(url_privileges, data=payload, headers=headers(content_type=content_type))
+        assert resp.status_code == 200, resp.status_code
+
+        # ensure user that creates the api keys has all privileges
+        # see how application privileges can be added to user roles in testing/docker/elasticsearch/roles.yml
+        url_user_privileges = "{}/_security/user/_has_privileges".format(cls.admin_es_url)
+        payload = json.dumps({"application": [{
+            "application": cls.application,
+            "privileges": [cls.privilege_config, cls.privilege_sourcemap, cls.privilege_intake],
+            "resources": cls.resource_backend}]})
+        resp = requests.post(url_user_privileges, data=payload, headers=headers(content_type=content_type))
+        assert resp.status_code == 200, resp.status_code
+        assert "has_all_requested" in resp.json(), resp.content
+        assert resp.json()["has_all_requested"] == True, resp.json()
+
+        cls.created_api_keys = []
```

Comment:
Why is this class level?  It appears to be deleted one at a time in `tearDownClass` - I would expect it as instance level and handled via `setUp` and `tearDown` instead

Reply:
The API Keys are immutable and do not need to change between the tests. Therefore I decided to create them once upfront, use in all tests and then destroy them again. 

What would be the advantage of creating/deleting on an instance level for this?